lib/ok_then/result/enum.ex

```diff
@@ -79,6 +79,74 @@ defmodule OkThen.Result.Enum do
     |> Result.tagged_map(tag, &Enum.reverse/1)
   end
 
+  @doc """
+  Modifies an Enum of results by applying `filter_function` to each group of values separately,
+  according to their tag.
+
+  **Note:** The ordering of results is not maintained.
+
+  ## Examples
+
+      iex> [{:ok, 1}, {:ok, 2}, :error, :error]
+      ...> |> Result.Enum.map_grouped_by_tag(fn
+      ...>   :ok, values -> Enum.map(values, &(&1 + 1))
+      ...>   :error, _values -> []
+      ...> end)
+      [{:ok, 2}, {:ok, 3}]
+
+      iex> [{:ok, 1}]
+      ...> |> Result.Enum.map_grouped_by_tag(fn
+      ...>   :ok, _values -> nil
+      ...> end)
+      ** (ArgumentError) Expected filter_function clause for tag :ok to return a list, but got: nil
+
+      iex> [{:ok, 1}, {:error, 1}, {:ok, 2}, {:error, 2}, :none]
+      ...> |> Result.Enum.map_grouped_by_tag(fn
+      ...>   :ok, values -> Enum.map(values, &(&1 + 1))
+      ...>   :error, values -> Enum.take(values, 1)
+      ...>   :none, _values -> []
+      ...> end)
+      [{:error, 1}, {:ok, 2}, {:ok, 3}]
+
+      iex> [{:some, 1}, :other, {:some, 2}]
+      ...> |> Result.Enum.map_grouped_by_tag(fn
+      ...>   :some, values -> Enum.map(values, &(&1 + 1))
+      ...>   :other, _values -> []
+      ...> end)
+      [{:some, 2}, {:some, 3}]
+
+      iex> [:ok, "hello", {:error, "hello"}, {1, 2}]
+      ...> |> Result.Enum.map_grouped_by_tag(fn
+      ...>   tag, _values when tag in [:ok, :error] -> []
+      ...>   :untagged, values -> values
+      ...> end)
+      [{:untagged, "hello"}, {:untagged, {1, 2}}]
+
+      iex> []
+      ...> |> Result.Enum.map_grouped_by_tag(fn
+      ...>   :ok, values -> Enum.map(values, &(&1 + 1))
+      ...> end)
+      []
+  """
+  @spec map_grouped_by_tag([Result.tagged()], (atom(), [any()] -> [any()])) :: [Result.tagged()]
+  def map_grouped_by_tag(results, filter_function)
+      when is_list(results) and is_function(filter_function, 2) do
+    results
+    |> group_by_tag()
+    |> Enum.flat_map(fn {tag, values} ->
+      filter_function.(tag, values)
+      |> case do
+        list when is_list(list) ->
+          Enum.map(list, &{tag, &1})
+
+        other ->
+          raise ArgumentError,
+                "Expected filter_function clause for tag #{Kernel.inspect(tag)} " <>
```

Comment:
The formatting is a little janky here, but it's technically valid so 🤷🏽

Reply:
Yeah, it's not brilliant. Can you think of a way to improve it?

Reply:
In Rust I would use `format!`, but I don't think Elixir has an equivalent

Reply:
I can't think of a better way without introducing a different type of janky